services/payments/src/simcore_service_payments/db/payments_transactions_repo.py

```diff
@@ -1,9 +1,190 @@
+#
+
+
+class PaymentsValueError(PydanticErrorMixin, ValueError):
+    msg_template = "Error in payment transaction '{payment_id}'"
+
+
+class PaymentNotFoundError(PaymentsValueError):
+    msg_template = "Payment transaction '{payment_id}' was not initialized."
+
+
+class PaymentAlreadyAckedError(PaymentsValueError):
     ...
+
+
+class PaymentAlreadyExistsError(PaymentsValueError):
+    msg_template = "Payment transaction '{payment_id}' was already initialized."
+
+
+#
+# base
+#
+
+
+class BaseRepository:
+    """
+    Repositories are pulled at every request
+    """
+
+    def __init__(self, db_engine: AsyncEngine):
+        assert db_engine is not None  # nosec
+        self.db_engine = db_engine
+
+
+#
+# repo
+#
+
+
+class PaymentsTransactionsRepo(BaseRepository):
+    async def insert_init_payment_transaction(
+        self,
+        payment_id: PaymentID,
+        price_dollars: Decimal,
+        osparc_credits: Decimal,
+        product_name: str,
+        user_id: UserID,
+        user_email: str,
+        wallet_id: WalletID,
+        comment: str | None,
+        initiated_at: datetime.datetime,
+    ) -> PaymentID:
+        """Annotates init-payment transaction"""
+        async with self.db_engine.begin() as conn:
+            await conn.execute(
+                payments_transactions.insert().values(
+                    payment_id=f"{payment_id}",
+                    price_dollars=price_dollars,
+                    osparc_credits=osparc_credits,
+                    product_name=product_name,
+                    user_id=user_id,
+                    user_email=user_email,
+                    wallet_id=wallet_id,
+                    comment=comment,
+                    initiated_at=initiated_at,
+                )
+            )
+            return payment_id
+
+    async def update_ack_payment_transaction(
+        self,
+        payment_id: PaymentID,
+        completion_state: PaymentTransactionState,
+        state_message: str | None,
+        invoice_url: HttpUrl,
+    ) -> PaymentsTransactionsDB:
+        """
+        - ACKs payment by updating state with SUCCESS, ...
+
+        Raises:
+            ValueError
+            PaymentNotFoundError
+            PaymentAlreadyAckedError
+
+        """
+
+        if completion_state == PaymentTransactionState.PENDING:
+            msg = f"cannot update state with {completion_state=} since it is already initiated"
+            raise ValueError(msg)
+
+        optional = {}
+        if state_message:
+            optional["state_message"] = state_message
+
+        async with self.db_engine.begin() as connection:
+            row = (
+                await connection.execute(
+                    sa.select(
+                        payments_transactions.c.initiated_at,
+                        payments_transactions.c.completed_at,
+                    )
+                    .where(payments_transactions.c.payment_id == f"{payment_id}")
+                    .with_for_update()
```

Comment:
Q: Cool, if I understand correctly the the selected row in this table is locked, until the whole transaction finishes?

Reply:
yes